### PR TITLE
client: do not timeout when wait is true

### PR DESCRIPTION
**Pull request by gyuho**
Current V2 watch waits by encoding URL with wait=true.
When a client sets 'no-sync', it requests directly to
proxy and the proxy redirects it by cloning the request
object, which leads to cancel the original request when
it times out and the cloned request gets closed prematurely.

This fixes https://github.com/coreos/etcd/issues/3894 by querying
the original client request in order to not use context timeout
when 'wait=true'.

### CODE REVIEW

**Comment by heyitsanthony** [comment]
would it be worthwhile to have an e2e test for this / proxying?

**Comment by gyuho** [comment]
@heyitsanthony Good idea. Will add e2e test for this case. Thanks,

File: client/client.go [hunk 1]
```diff
@@ -442,9 +443,16 @@ func (c *simpleHTTPClient) Do(ctx context.Context, act httpAction) (*http.Respon
 		return nil, nil, err
 	}
 
+	isWait := false
+	if req != nil {
+		if req.URL != nil {
```
Comment (heyitsanthony) [inline client/client.go]: `&&` instead of nesting if's

Comment (gyuho) [inline client/client.go]: ok will fix. thanks

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
@heyitsanthony Just added e2e test that only fails without this patch.

```
--- FAIL: TestBasicOpsV2CtlWatchWithProxyNoSync (6.91s)
	etcd_test.go:190: failed watch (Error:  client: etcd cluster is unav)ilable or misconfigured
FAIL
exit status 1
```

With this patch, e2e test passes:

PTAL. Thanks!

File: e2e/etcd_test.go [hunk 2]
```diff
@@ -42,6 +44,7 @@ func TestBasicOpsNoTLS(t *testing.T) {
 			isPeerTLS:    false,
 			initialToken: "new",
 		},
+		etcdProcessBasePort,
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: why have a new base port argument? the same ports were reusable across tests

File: client/client.go [hunk 3]
```diff
@@ -442,9 +443,14 @@ func (c *simpleHTTPClient) Do(ctx context.Context, act httpAction) (*http.Respon
...
 
+	isWait := false
+	if req != nil && req.URL != nil {
+		isWait, _ = strconv.ParseBool(req.URL.Query().Get("wait"))
+	}
+
 	var hctx context.Context
 	var hcancel context.CancelFunc
-	if c.headerTimeout > 0 {
+	if !isWait && c.headerTimeout > 0 {
```
Comment (xiang90) [inline client/client.go]: @gyuho 

This does not seem to be a 100% correct change. Even for the watch case, the header should be returned first before the body which is to wait on. I suspect that our etcd proxy does not forward header in time though. Not sure about other cases of proxy. We have to make a tradeoff here.

File: e2e/etcd_test.go (hunk 2 above)
Comment (gyuho) [inline e2e/etcd_test.go]: It seemed like test was hanging for watch cases when using the same ports.

File: client/client.go (hunk 3 above)
Comment (gyuho) [inline client/client.go]: @xiang90 It copies header https://github.com/coreos/etcd/blob/master/proxy/reverse.go#L89 here, but do you think it does something else wrong?

File: e2e/etcd_test.go [hunk 4]
```diff
@@ -103,6 +109,108 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
...
+		if errC := epc.Close(); errC != nil {
+			t.Fatalf("error closing etcd processes (%v)", errC)
+		}
+	}()
+
+	endpoint, be := "", ""
+	for _, p := range epc.procs {
+		if p.cfg.isProxy {
+			endpoint = p.cfg.acurl.String()
+			time.Sleep(5 * time.Second) // give some time for proxy to find cluster
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: this is racey; instead, `newEtcdProcessCluster` shouldn't return until the proxies are ready

File: client/client.go (hunk 3 above)
Comment (xiang90) [inline client/client.go]: It copies but not flush in time. Copy only means put into memory I guess. But again, even if we fix our proxy, there are other proxies that do not flush header in time.

File: e2e/etcd_test.go (hunk 4 above)
Comment (gyuho) [inline e2e/etcd_test.go]: Ok, will sync before return, if @xiang90  approves.

File: client/client.go [hunk 5]
```diff
@@ -442,9 +443,14 @@ func (c *simpleHTTPClient) Do(ctx context.Context, act httpAction) (*http.Respon
 		return nil, nil, err
 	}
 
+	isWait := false
+	if req != nil && req.URL != nil {
+		isWait, _ = strconv.ParseBool(req.URL.Query().Get("wait"))
```
Comment (xiang90) [inline client/client.go]: do not swallow error please. i assume this is an error. I am having hard time to use godoc at the memory at China...

Comment (gyuho) [inline client/client.go]: Yes it's error. I will check error here.

File: e2e/etcd_test.go [hunk 6]
```diff
@@ -103,6 +109,108 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
...
+			t.Fatalf("failed watch (%v)", err)
+		}
+
+		s, _ := proc.ReadLine()
+		if strings.Contains(s, "client: etcd cluster is unavailable or misconfigured") {
+			t.Fatalf("failed watch (%v)", s)
+		}
+	}()
+
+	time.Sleep(time.Second)
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: why the `Sleep`?

Comment (gyuho) [inline e2e/etcd_test.go]: Will delete. Had it before I was using wait group. Thanks,

File: client/client.go (hunk 3 above)
Comment (gyuho) [inline client/client.go]: @xiang90 if I understand your comment correctly, we can flush the buffer for each request like

```go
if f, ok := rw.(http.Flusher); ok {
    if f != nil {
      f.Flush()
    }
}
```

but this won't fix the root cause when there are more than 1 proxy?

If so, what do you suggest to handle such cases? I won't proceed with this PR
until we figure out the best approach.

Thanks for feedback!

Comment (xiang90) [inline client/client.go]: Please check the comment here https://github.com/coreos/etcd/blob/master/client/client.go#L123-L124.

Basically, the requestHeader behavior is undefined when there is a proxy in front of the client. 

You are changing the default behavior from applying the timeout to not apply it. This change seems to be OK with me. Even if we "fix" etcd proxy, we will not be able to control all the proxies people are using. 

But we should make the document better by saying for wait request we will ignore this timeout.

**Comment by xiang90** [comment]
@gyuho This change seems to be OK. But please update the documentation accordingly. 

defer to @heyitsanthony on the testing changes.

**Comment by gyuho** [comment]
@xiang90 Now understand what you were talking about. I will make the comment clearer. And will update tests addressing Anthony's feedback.

Thanks,

File: e2e/etcd_test.go [hunk 7]
```diff
@@ -103,6 +109,108 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
...
+	if noSync {
+		args = append(args, "--no-sync")
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		args = append(args, "watch", key)
+		proc, err := spawnCmd(args)
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: this needs a `proc.Close()` (`spawnWithExpect` will do it for you)

File: e2e/etcd_test.go [hunk 8]
```diff
@@ -103,6 +109,108 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
...
+	}()
+
+	time.Sleep(time.Second)
+
+	putArgs := []string{"../bin/etcdctl", "--endpoint", endpoint}
+	if noSync {
+		args = append(args, "--no-sync")
+	}
+	putArgs = append(putArgs, "set", key, value)
+	proc, err := spawnCmd(putArgs)
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: also needs a `proc.Close()`

File: e2e/etcd_test.go (hunk 2 above)
Comment (heyitsanthony) [inline e2e/etcd_test.go]: sounds broken; a process might be hanging

**Comment by xiang90** [comment]
@gyuho @heyitsanthony 

Basically the whole long polling behavior is broken for the proxy case when proxy does not forward response header or response in time. The request header is designed for watch request in the first place. Or we have no way to know if the remote machine is hang or not. I am not even sure if we still need request header timeout or not.

File: e2e/etcd_test.go [hunk 9]
```diff
@@ -103,6 +109,108 @@ func testProcessClusterPutGet(t *testing.T, cfg *etcdProcessClusterConfig) {
...
+		t.Fatalf("could not start etcd process cluster (%v)", errC)
+	}
+	defer func() {
+		if errC := epc.Close(); errC != nil {
+			t.Fatalf("error closing etcd processes (%v)", errC)
+		}
+	}()
+
+	endpoint, be := "", ""
+	for _, p := range epc.procs {
```
Comment (heyitsanthony) [inline e2e/etcd_test.go]: this loop should probably be replaced by calling methods like `(e *EtcdProcessCluster) Members() []*etcdProcess` and `(e *EtcdProcessCluster) Proxies() []*etcdProcess`

**Comment by gyuho** [comment]
I think `e2e` test is currently broken. https://github.com/coreos/etcd/pull/4257
I locally tested the `e2e` and confirmed that this fixes the issue https://github.com/coreos/etcd/issues/3894.

Let's fix the `e2e` issue in a separate PR.
Once we figure out the goroutine leak issue, I will rewrite the test.

**Comment by gyuho** [comment]
@heyitsanthony Could you take a look at `client.go` again? I will write the end to end test in a separate PR. Need to figure out why it was hanging. Thanks!

**Comment by heyitsanthony** [comment]
lgtm

**Comment by gyuho** [comment]
Thanks!

**Pull request merged**